src/subscription_manager/syspurposelib.py

```diff
@@ -216,5 +239,146 @@ def read_syspurpose():
             syspurpose = json.load(open(USER_SYSPURPOSE))
         except (os.error, ValueError):
             # In the event this file could not be read treat it as empty
+            if raise_on_error:
+                raise
             syspurpose = {}
     return syspurpose
+
+
+def write_syspurpose(values):
+    """
+    Write the syspurpose to the file system.
+    :param values:
+    :return:
+    """
+    if SyspurposeStore is not None:
+        sp = SyspurposeStore(USER_SYSPURPOSE)
+        sp.contents = values
+        sp.write()
+    else:
+        # Simple backup in case the syspurpose tooling is not installed.
+        try:
+            json.dump(values, open(USER_SYSPURPOSE), ensure_ascii=True, indent=2)
+        except OSError:
+            log.warning('Could not write syspurpose to %s' % USER_SYSPURPOSE)
+            return False
+    return True
+
+
+class SyspurposeSyncActionInvoker(certlib.BaseActionInvoker):
+    """
+    Used by rhsmcertd to sync the syspurpose values locally with those from the Server.
+    """
+
+    def _do_update(self):
+        action = SyspurposeSyncActionCommand()
+        return action.perform()
+
+
+class SyspurposeSyncActionReport(certlib.ActionReport):
+    name = "Syspurpose Sync"
+
+    def record_change(self, change):
+        """
+        Records the change detected by the three_way_merge function into a record in the report.
+        :param change: A util.DiffChange object containing the recorded changes.
+        :return: None
+        """
+        if change.source == 'remote':
+            source = 'Entitlement Server'
+        elif change.source == 'local':
+            source = USER_SYSPURPOSE
+        else:
+            source = 'cached system purpose values'
+        msg = None
+        if change.in_base and not change.in_result:
+            msg = "'{key}' removed by change from {source}".format(key=change.key,
+                                                                     source=source)
+        elif not change.in_base and change.in_result:
+            msg = "'{key}' added with value '{value}' from change in {source}".format(
+                    key=change.key, value=change.new_value, source=source
+            )
+        elif change.in_base and change.previous_value != change.new_value:
+            msg = "'{key}' updated to '{new_value}' from '{old_value}' from change in {source}"\
```

Comment:
This message could be little bit confusing. I would use here something like this:
```python
"'{key}' updated from '{old_value}' to '{new_value}' due to change in {source}"
```